envoy/ssl/handshaker.h

```diff
@@ -156,5 +165,96 @@ class HandshakerFactory : public Config::TypedFactory {
   virtual SslCtxCb sslctxCb(HandshakerFactoryContext& handshaker_factory_context) const PURE;
 };
 
+enum class SelectionResult {
+  // continue the TLS handshake.
+  Continue,
+  // block the TLS handshake.
+  Stop,
+  // terminate the TLS handshake.
+  Terminate
+};
+
+/**
+ * Used to return the result from an synchronous/asynchronous cert selection.
+ */
+class CertSelectionCallback {
+public:
+  virtual ~CertSelectionCallback() = default;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * Called when the cert selection completes.
+   * It's safe to call it even the SSL connection may be terminated early.
+   */
+  virtual void onCertSelectionResult(bool succeeded, const Ssl::TlsContext& selected_ctx,
+                                     bool staple) PURE;
+};
+
+using CertSelectionCallbackSharedPtr = std::shared_ptr<CertSelectionCallback>;
```

Comment:
I had tried `std::unique_ptr`, similar to `ValidateResultCallbackImpl`, but leaks to CI failure.

Here is the reason:
while we implementing a `TlsCertificateSelector` extension, then the extension owner the `CertSelectionCallbackPtr cb`.
such as the test case:
https://github.com/envoyproxy/envoy/blob/fcd2f3c2ecf318d0678f02c3c122e7a7013e60f6/test/common/tls/tls_certificate_selector_test.cc#L90

but, this could be happens:
1. when `selectTlsContext` returns with `SelectionResult::Stop`, and the extension do not save the `cb`, then the `cb` will be freed.
2. `cert_selection_callback_` will be an invalid pointer then.
3. when the tls connection is closed, may closed from client, segfault will happens in:

https://github.com/envoyproxy/envoy/blob/fcd2f3c2ecf318d0678f02c3c122e7a7013e60f6/source/common/tls/ssl_handshaker.cc#L65-L67

`std::shared_ptr` might be a proper choice?